File: src/main/java/io/dev/mutex/thesisinfomgmt/service/RoleService.java
package io.dev.mutex.thesisinfomgmt.service;

import io.dev.mutex.thesisinfomgmt.common.PaginatedData;
import io.dev.mutex.thesisinfomgmt.dto.RoleDTO;

public interface RoleService {

  /**
   * Retrieves roles
   *
   * @param page     the page of the data to be retrieved
   * @param pageSize the number of items to be retrieved
   * @param query    the query string
   * @return paginated roles data
   */
  PaginatedData<RoleDTO> getRoles(int page, int pageSize, String query);

  /**
   * Creates a new role
   *
   * @param role the details of the new role
   * @return create role
   */
  RoleDTO createRole(RoleDTO role);

  /**
   * Retrieves a role by its id
   *
   * @param id the id of the role to be retrieved
   * @return retrieved role
   */
  RoleDTO getRole(long id);

  /**
   * Deletes a role
   *
   * @param id the id of the role to be deleted
   */
  void deleteRole(long id);

  /**
   * Updates a role
   *
   * @param id   the id of the role to be updated
   * @param role the updated role details
   */
  void updateRole(long id, RoleDTO role);
}

File: app/src/services/RoleService.ts
import { PaginatedData } from "../types/Common";
import { Role } from "../types/Role";
import RequestHelper from "../helpers/RequestHelper";
import qs from "qs";
import { RequestMethod } from "../common/constants";

class RoleService {
  resourcePath = "roles";

  async getRoles(
    page: number,
    pageSize: number,
    query: string
  ): Promise<PaginatedData<Role>> {
    const queryParams = {
      page: page,
      pageSize: pageSize,
      query: query,
    };
    return await RequestHelper.doRemoteCallAsync(
      `${this.resourcePath}?${qs.stringify(queryParams)}`,
      "get roles"
    );
  }

  async createRole(role: Role): Promise<Role> {
    return await RequestHelper.doRemoteCallAsync(
      this.resourcePath,
      "create role",
      RequestMethod.POST,
      role
    );
  }

  async updateRole(roleId: number, role: Role): Promise<void> {
    return await RequestHelper.doRemoteCallAsync(
      `${this.resourcePath}/${roleId}`,
      "update role",
      RequestMethod.PUT,
      role
    );
  }

  async deleteRole(roleId: number): Promise<void> {
    return await RequestHelper.doRemoteCallAsync(
      `${this.resourcePath}/${roleId}`,
      "delete role",
      RequestMethod.DELETE
    );
  }
}

export default new RoleService();
